core/unix/coredump.c

```diff
@@ -0,0 +1,362 @@
+
+    return os_write(elf_file, (void *)&phdr, sizeof(ELF_PROGRAM_HEADER_TYPE)) ==
+        sizeof(ELF_PROGRAM_HEADER_TYPE);
+}
+
+/*
+ * Write a section header to the file. Returns true if the section header is written to
+ * the core dump file, false otherwise.
+ */
+static bool
+write_shdr(DR_PARAM_IN file_t elf_file, DR_PARAM_IN ELF_WORD string_table_offset,
+           DR_PARAM_IN ELF_WORD type, DR_PARAM_IN ELF_WORD flags,
+           DR_PARAM_IN ELF_ADDR virtual_address, DR_PARAM_IN ELF_OFF offset,
+           DR_PARAM_IN ELF_WORD section_size, DR_PARAM_IN ELF_WORD link,
+           DR_PARAM_IN ELF_WORD info, DR_PARAM_IN ELF_WORD alignment,
+           DR_PARAM_IN ELF_WORD entry_size)
+{
+    ELF_SECTION_HEADER_TYPE shdr;
+    shdr.sh_name = string_table_offset; /* Section name (string tbl index) */
+    shdr.sh_type = type;                /* Section type */
+    shdr.sh_flags = flags;              /* Section flags */
+    shdr.sh_addr = virtual_address;     /* Section virtual addr at execution */
+    shdr.sh_offset = offset;            /* Section file offset */
+    shdr.sh_size = section_size;        /* Section size in bytes */
+    shdr.sh_link = link;                /* Link to another section */
+    shdr.sh_info = info;                /* Additional section information */
+    shdr.sh_addralign = alignment;      /* Section alignment */
+    shdr.sh_entsize = entry_size;       /* Entry size if section holds table */
+
+    return os_write(elf_file, (void *)&shdr, sizeof(ELF_SECTION_HEADER_TYPE)) ==
+        sizeof(ELF_SECTION_HEADER_TYPE);
+}
+
+/*
+ * Writes a memory dump file in ELF format. Returns true if a core dump file is written,
+ * false otherwise.
+ */
+static bool
+os_dump_core_internal(void)
+{
+    strhash_table_t *string_htable =
+        strhash_hash_create(GLOBAL_DCONTEXT, /*bits=*/8, /*load_factor_percent=*/80,
+                            /*table_flags=*/0, NULL _IF_DEBUG("mmap-string-table"));
+
+    // Insert a null string at the beginning for sections with no comment.
+    char string_table[MAX_SECTION_NAME_BUFFER_SIZE];
+    string_table[0] = '\0';
+    string_table[1] = '\0';
+    ELF_ADDR string_table_offset = 1;
+    ELF_OFF section_count = 0;
+    ELF_OFF seciion_data_size = 0;
+    memquery_iter_t iter;
+    if (memquery_iterator_start(&iter, NULL, /*may_alloc=*/false)) {
+        while (memquery_iterator_next(&iter)) {
+            // Skip non-readable section.
+            if (iter.prot == MEMPROT_NONE || strcmp(iter.comment, VVAR_SECTION) == 0) {
+                continue;
+            }
+            ELF_ADDR offset = 0;
+            if (iter.comment != NULL && iter.comment[0] != '\0') {
+                offset = (ELF_ADDR)strhash_hash_lookup(GLOBAL_DCONTEXT, string_htable,
+                                                       iter.comment);
+                if (offset == 0 &&
+                    (string_table[1] == '\0' ||
+                     d_r_strcmp(string_table, iter.comment) != 0)) {
```

Comment:
I don't understand this: the start of string_table is 2 nul chars.  This will always fail.  Is this supposed to be strhash_hash_lookup?

Reply:
I have cleaned up the logic and used strhash_hash_lookup() to decide if the section name has been seen or not.
PTAL.